File: binom_assistant/modules/portfolio/total_performance_tracker.py
"""
Модуль отслеживания общей динамики портфеля (Total Performance Tracker)
"""
from typing import Dict, Any, List
from datetime import datetime, timedelta
from sqlalchemy import func
from contextlib import contextmanager
from collections import defaultdict

from storage.database.base import get_session
from storage.database.models import Campaign, CampaignStatsDaily
from ..base_module import BaseModule, ModuleMetadata, ModuleConfig


@contextmanager
def get_db_session():
    """
    Локальная обертка над get_session() для использования в with.
    Преобразует генератор в контекстный менеджер.
    """
    session_gen = get_session()
    session = next(session_gen)
    try:
        yield session
    finally:
        try:
            next(session_gen)
        except StopIteration:
            pass


class TotalPerformanceTracker(BaseModule):
    """
    Модуль отслеживания общей динамики портфеля (Total Performance Tracker).

    Отслеживает динамику общих показателей:
    - Агрегированные метрики: общий ROI, расход, прибыль
    - Сравнение периодов (текущий vs предыдущий)
    - Тренд изменения метрик
    """

    def get_metadata(self) -> ModuleMetadata:
        """Возвращает метаданные модуля"""
        return ModuleMetadata(
            id="total_performance_tracker",
            name="Общая динамика",
            category="portfolio",
            description="Отслеживает динамику общих показателей портфеля",
            detailed_description="Модуль анализирует общие показатели портфеля за выбранный период и сравнивает их с предыдущим периодом. Показывает тренд изменения ключевых метрик: ROI, расход, прибыль, клики, лиды. Помогает быстро оценить общую динамику портфеля. Анализирует только полные дни, исключая сегодняшний (апрувы приходят с задержкой).",
            version="1.1.0",
            author="Binom Assistant",
            priority="high",
            tags=["portfolio", "performance", "dynamics", "trends", "roi", "metrics"]
        )

    def get_default_config(self) -> ModuleConfig:
        """Возвращает конфигурацию по умолчанию"""
        return ModuleConfig(
            enabled=True,
            schedule="",  # Некритический модуль - не запускать автоматически
            alerts_enabled=False,  # Алерты выключены по умолчанию
            timeout_seconds=60,
            cache_ttl_seconds=3600,
            params={
                "days": 7,  # период для анализа
                "min_change_threshold": 2.0,  # минимальное изменение для уведомлений (%)
                "high_roi_threshold": 20,  # порог высокого ROI (%)
                "high_cost_change_threshold": 20  # порог значительного изменения расходов (%)
            }
        )

    def get_param_metadata(self) -> Dict[str, Dict[str, Any]]:
        """Возвращает метаданные параметров для UI"""
        return {
            "days": {
                "label": "Период анализа (дней)",
                "description": "Количество дней для анализа портфеля",
                "type": "number",
                "min": 1,
                "max": 365,
                "default": 7
            },
            "min_change_threshold": {
                "label": "Порог изменения (%)",
                "description": "Минимальное изменение метрики для генерации уведомления",
                "type": "number",
                "min": 0.1,
                "max": 50,
                "default": 2.0
            },
            "high_roi_threshold": {
                "label": "Порог высокого ROI (%)",
                "description": "Порог для определения высокого ROI",
                "type": "number",
                "min": 0,
                "max": 200,
                "default": 20
            },
            "high_cost_change_threshold": {
                "label": "Порог изменения расходов (%)",
                "description": "Порог значительного изменения расходов",
                "type": "number",
                "min": 5,
                "max": 10000,
                "default": 20
            }
        }

    def analyze(self, config: ModuleConfig) -> Dict[str, Any]:
        """
        Анализ общей динамики портфеля.

        Args:
            config: Конфигурация модуля

        Returns:
            Dict[str, Any]: Данные о динамике портфеля
        """
        # Получение параметров
        days = config.params.get("days", 7)
        min_change_threshold = config.params.get("min_change_threshold", 2.0)
        high_roi_threshold = config.params.get("high_roi_threshold", 20)
        high_cost_change_threshold = config.params.get("high_cost_change_threshold", 20)

        # Исключаем сегодняшний день (апрувы приходят с задержкой)
        date_from = datetime.now().date() - timedelta(days=days)
        date_to = datetime.now().date()

        # Предыдущий период (для сравнения)
        prev_date_from = date_from - timedelta(days=days)
        prev_date_to = date_from - timedelta(days=1)

        # Работа с БД
        with get_db_session() as session:
            # Получаем данные за текущий период
            current_data = self._aggregate_period_data(
                session, date_from, date_to
            )

            # Получаем данные за предыдущий период
            previous_data = self._aggregate_period_data(
                session, prev_date_from, prev_date_to
            )

            # Расчет изменений
            changes = self._calculate_changes(current_data, previous_data)

            # Определение тренда
            trend = self._determine_trend(changes, min_change_threshold)

            # Формируем summary для отображения в карточке модуля
            summary = {
                "total_cost": current_data.get("total_cost", 0),
                "total_revenue": current_data.get("total_revenue", 0),
                "total_profit": current_data.get("total_profit", 0),
                "roi": current_data.get("total_roi", 0),
                "clicks": current_data.get("total_clicks", 0),
                "leads": current_data.get("total_leads", 0),
                "approved_leads": current_data.get("total_a_leads", 0),
                "trend": trend
            }

            return {
                "summary": summary,
                "current_period": current_data,
                "previous_period": previous_data,
                "changes": changes,
                "trend": trend,
                "period": {
                    "days": days,
                    "date_from": date_from.isoformat(),
                    "date_to": date_to.isoformat()
                }
            }

    def _aggregate_period_data(self, session, date_from, date_to) -> Dict[str, Any]:
        """
        Агрегирует данные за период.

        Args:
            session: Сессия БД
            date_from: Начальная дата
            date_to: Конечная дата

        Returns:
            Dict[str, Any]: Агрегированные данные за период
        """
        query = session.query(
            func.sum(CampaignStatsDaily.cost).label('total_cost'),
            func.sum(CampaignStatsDaily.revenue).label('total_revenue'),
            func.sum(CampaignStatsDaily.leads).label('total_leads'),
            func.sum(CampaignStatsDaily.a_leads).label('total_a_leads'),
            func.sum(CampaignStatsDaily.clicks).label('total_clicks')
        ).filter(
            CampaignStatsDaily.date >= date_from,
            CampaignStatsDaily.date <= date_to
        )

        result = query.first()

        # Обработка None значений
        total_cost = float(result.total_cost or 0)
        total_revenue = float(result.total_revenue or 0)
        total_leads = float(result.total_leads or 0)
        total_a_leads = float(result.total_a_leads or 0)
        total_clicks = float(result.total_clicks or 0)

        # Расчет ROI
        total_roi = ((total_revenue - total_cost) / total_cost * 100) if total_cost > 0 else 0

        return {
            "total_cost": round(total_cost, 2),
            "total_revenue": round(total_revenue, 2),
            "total_profit": round(total_revenue - total_cost, 2),
            "total_roi": round(total_roi, 1),
            "total_clicks": int(total_clicks),
            "total_leads": int(total_leads),
            "total_a_leads": int(total_a_leads)
        }

    def _calculate_changes(self, current: Dict[str, Any], previous: Dict[str, Any]) -> Dict[str, float]:
        """
        Рассчитывает процентные изменения между периодами.

        Args:
            current: Текущие данные
            previous: Предыдущие данные

        Returns:
            Dict[str, float]: Процентные изменения
        """
        changes = {}

        # Изменение расхода
        if previous['total_cost'] > 0:
            changes['cost_change'] = round(
                ((current['total_cost'] - previous['total_cost']) / previous['total_cost']) * 100,
                1
            )
        else:
            changes['cost_change'] = 0 if current['total_cost'] == 0 else 100

        # Изменение дохода
        if previous['total_revenue'] > 0:
            changes['revenue_change'] = round(
                ((current['total_revenue'] - previous['total_revenue']) / previous['total_revenue']) * 100,
                1
            )
        else:
            changes['revenue_change'] = 0 if current['total_revenue'] == 0 else 100

        # Изменение прибыли
        if previous['total_profit'] > 0:
            changes['profit_change'] = round(
                ((current['total_profit'] - previous['total_profit']) / previous['total_profit']) * 100,
                1
            )
        elif previous['total_profit'] == 0 and current['total_profit'] > 0:
            changes['profit_change'] = 100
        elif previous['total_profit'] == 0 and current['total_profit'] == 0:
            changes['profit_change'] = 0
        else:
            changes['profit_change'] = -100

        # Изменение ROI
        if previous['total_roi'] != 0:
            changes['roi_change'] = round(
                current['total_roi'] - previous['total_roi'],
                1
            )
        else:
            changes['roi_change'] = current['total_roi'] - 0

        # Изменение кликов
        if previous['total_clicks'] > 0:
            changes['clicks_change'] = round(
                ((current['total_clicks'] - previous['total_clicks']) / previous['total_clicks']) * 100,
                1
            )
        else:
            changes['clicks_change'] = 0 if current['total_clicks'] == 0 else 100

        # Изменение лидов
        if previous['total_leads'] > 0:
            changes['leads_change'] = round(
                ((current['total_leads'] - previous['total_leads']) / previous['total_leads']) * 100,
                1
            )
        else:
            changes['leads_change'] = 0 if current['total_leads'] == 0 else 100

        # Изменение одобренных лидов
        if previous['total_a_leads'] > 0:
            changes['a_leads_change'] = round(
                ((current['total_a_leads'] - previous['total_a_leads']) / previous['total_a_leads']) * 100,
                1
            )
        else:
            changes['a_leads_change'] = 0 if current['total_a_leads'] == 0 else 100

        return changes

    def _determine_trend(self, changes: Dict[str, float], threshold: float = 2.0) -> str:
        """
        Определяет общий тренд на основе изменений.

        Args:
            changes: Словарь с процентными изменениями
            threshold: Пороговое значение для определения значимых изменений (%)

        Returns:
            str: Тренд ("improving", "stable", "declining")
        """
        # Ключевые метрики для определения тренда
        revenue_change = changes.get('revenue_change', 0)
        roi_change = changes.get('roi_change', 0)
        profit_change = changes.get('profit_change', 0)

        # Подсчитываем положительные и отрицательные изменения
        positive_count = 0
        negative_count = 0

        if revenue_change > threshold:
            positive_count += 1
        elif revenue_change < -threshold:
            negative_count += 1

        if roi_change > threshold:
            positive_count += 1
        elif roi_change < -threshold:
            negative_count += 1

        if profit_change > threshold:
            positive_count += 1
        elif profit_change < -threshold:
            negative_count += 1

        # Определяем тренд
        if positive_count > negative_count:
            return "improving"
        elif negative_count > positive_count:
            return "declining"
        else:
            return "stable"

    def generate_recommendations(self, raw_data: Dict[str, Any]) -> List[str]:
        """
        Генерация рекомендаций - ОТКЛЮЧЕНО для этого модуля.
        """
        return []

    def prepare_chart_data(self, raw_data: Dict[str, Any]) -> List[Dict[str, Any]]:
        """Подготовка данных для Chart.js"""
        current = raw_data.get("current_period", {})
        previous = raw_data.get("previous_period", {})
        changes = raw_data.get("changes", {})

        charts = []

        # График сравнения ключевых метрик
        charts.append({
            "id": "total_performance_comparison",
            "type": "bar",
            "data": {
                "labels": ["Текущий период", "Предыдущий период"],
                "datasets": [
                    {
                        "label": "Расход ($)",
                        "data": [
                            current.get('total_cost', 0),
                            previous.get('total_cost', 0)
                        ],
                        "backgroundColor": "rgba(255, 159, 64, 0.7)"
                    },
                    {
                        "label": "Доход ($)",
                        "data": [
                            current.get('total_revenue', 0),
                            previous.get('total_revenue', 0)
                        ],
                        "backgroundColor": "rgba(75, 192, 75, 0.7)"
                    },
                    {
                        "label": "Прибыль ($)",
                        "data": [
                            current.get('total_profit', 0),
                            previous.get('total_profit', 0)
                        ],
                        "backgroundColor": "rgba(54, 162, 235, 0.7)"
                    }
                ]
            },
            "options": {
                "responsive": True,
                "plugins": {
                    "title": {
                        "display": True,
                        "text": "Сравнение периодов"
                    },
                    "legend": {
                        "position": "bottom"
                    }
                },
                "scales": {
                    "y": {
                        "beginAtZero": True
                    }
                }
            }
        })

        # График изменений в процентах
        roi_color = "rgba(40, 167, 69, 0.7)" if changes.get('roi_change', 0) >= 0 else "rgba(220, 53, 69, 0.7)"
        revenue_color = "rgba(40, 167, 69, 0.7)" if changes.get('revenue_change', 0) >= 0 else "rgba(220, 53, 69, 0.7)"
        profit_color = "rgba(40, 167, 69, 0.7)" if changes.get('profit_change', 0) >= 0 else "rgba(220, 53, 69, 0.7)"

        charts.append({
            "id": "total_performance_changes",
            "type": "bar",
            "data": {
                "labels": ["Расход", "Доход", "Прибыль", "ROI", "Клики", "Лиды"],
                "datasets": [{
                    "label": "Изменение (%)",
                    "data": [
                        changes.get('cost_change', 0),
                        changes.get('revenue_change', 0),
                        changes.get('profit_change', 0),
                        changes.get('roi_change', 0),
                        changes.get('clicks_change', 0),
                        changes.get('leads_change', 0)
                    ],
                    "backgroundColor": [
                        "rgba(255, 159, 64, 0.7)",  # cost
                        "rgba(40, 167, 69, 0.7)" if changes.get('revenue_change', 0) >= 0 else "rgba(220, 53, 69, 0.7)",
                        "rgba(40, 167, 69, 0.7)" if changes.get('profit_change', 0) >= 0 else "rgba(220, 53, 69, 0.7)",
                        "rgba(40, 167, 69, 0.7)" if changes.get('roi_change', 0) >= 0 else "rgba(220, 53, 69, 0.7)",
                        "rgba(40, 167, 69, 0.7)" if changes.get('clicks_change', 0) >= 0 else "rgba(220, 53, 69, 0.7)",
                        "rgba(40, 167, 69, 0.7)" if changes.get('leads_change', 0) >= 0 else "rgba(220, 53, 69, 0.7)"
                    ]
                }]
            },
            "options": {
                "responsive": True,
                "plugins": {
                    "title": {
                        "display": True,
                        "text": "Изменение метрик vs предыдущий период"
                    },
                    "legend": {
                        "position": "bottom"
                    }
                },
                "scales": {
                    "y": {
                        "title": {
                            "display": True,
                            "text": "Процентное изменение (%)"
                        }
                    }
                }
            }
        })

        # График ROI за периоды
        charts.append({
            "id": "total_performance_roi",
            "type": "line",
            "data": {
                "labels": ["Предыдущий период", "Текущий период"],
                "datasets": [{
                    "label": "ROI (%)",
                    "data": [
                        previous.get('total_roi', 0),
                        current.get('total_roi', 0)
                    ],
                    "borderColor": "rgba(13, 110, 253, 1)",
                    "backgroundColor": "rgba(13, 110, 253, 0.1)",
                    "borderWidth": 2,
                    "pointBackgroundColor": "rgba(13, 110, 253, 1)",
                    "pointBorderColor": "#fff",
                    "pointBorderWidth": 2,
                    "tension": 0.4
                }]
            },
            "options": {
                "responsive": True,
                "plugins": {
                    "title": {
                        "display": True,
                        "text": "Динамика ROI"
                    },
                    "legend": {
                        "position": "bottom"
                    }
                },
                "scales": {
                    "y": {
                        "title": {
                            "display": True,
                            "text": "ROI (%)"
                        }
                    }
                }
            }
        })

        return charts

    def generate_alerts(self, raw_data: Dict[str, Any]) -> List[Dict[str, Any]]:
        """
        Генерация алертов для портфеля.
        """
        current = raw_data.get("current_period", {})
        previous = raw_data.get("previous_period", {})
        changes = raw_data.get("changes", {})
        trend = raw_data.get("trend", "stable")

        alerts = []

        # Алерт о тренде
        trend_message = {
            "improving": "Портфель улучшается! Все ключевые метрики растут.",
            "stable": "Портфель стабилен. Метрики остаются на прежнем уровне.",
            "declining": "Внимание! Портфель ухудшается. Требуется анализ причин."
        }.get(trend, "Неизвестный тренд")

        severity = {
            "improving": "info",
            "stable": "info",
            "declining": "warning"
        }.get(trend, "info")

        alerts.append({
            "type": "performance_trend",
            "severity": severity,
            "message": trend_message,
            "trend": trend
        })

        # Алерт о ROI
        roi_current = current.get('total_roi', 0)
        roi_change = changes.get('roi_change', 0)

        if roi_current > 50:
            roi_message = f"Отличный ROI: {roi_current}% (+{roi_change}%)"
            roi_severity = "info"
        elif roi_current > 20:
            roi_message = f"Хороший ROI: {roi_current}% ({roi_change:+.1f}%)"
            roi_severity = "info"
        elif roi_current > 0:
            roi_message = f"ROI положительный: {roi_current}% ({roi_change:+.1f}%)"
            roi_severity = "warning"
        else:
            roi_message = f"ROI отрицательный: {roi_current}% ({roi_change:+.1f}%)"
            roi_severity = "critical"

        alerts.append({
            "type": "performance_roi",
            "severity": roi_severity,
            "message": roi_message,
            "roi": roi_current,
            "roi_change": roi_change
        })

        # Алерт о расходах
        cost_change = changes.get('cost_change', 0)
        if cost_change > 20:
            cost_message = f"Резкий рост расходов: +{cost_change}%. Убедитесь, что бюджет контролируется."
            cost_severity = "warning"
        elif cost_change < -20:
            cost_message = f"Значительное снижение расходов: {cost_change}%. Проверьте, не заблокированы ли кампании."
            cost_severity = "info"
        else:
            cost_message = f"Расходы изменились на {cost_change:+.1f}%"
            cost_severity = "info"

        alerts.append({
            "type": "performance_cost",
            "severity": cost_severity,
            "message": cost_message,
            "cost_change": cost_change
        })

        # Алерт о прибыли
        profit_current = current.get('total_profit', 0)
        profit_change = changes.get('profit_change', 0)

        if profit_current > 0:
            profit_status = f"Положительная прибыль: ${profit_current:.2f}"
            if profit_change > 0:
                profit_status += f" (+{profit_change}%)"
            else:
                profit_status += f" ({profit_change}%)"
            profit_severity = "info"
        else:
            profit_status = f"Отрицательная прибыль: ${profit_current:.2f} ({profit_change:+.1f}%)"
            profit_severity = "critical"

        alerts.append({
            "type": "performance_profit",
            "severity": profit_severity,
            "message": profit_status,
            "profit": profit_current,
            "profit_change": profit_change
        })

        return alerts

File: binom_assistant/interfaces/web/static/js/modules/total_performance_tracker.js
/**
 * Модуль: Общая динамика портфеля (Total Performance Tracker)
 * Отслеживает динамику общих показателей портфеля
 */
(function() {
    const TotalPerformanceTrackerModule = {
        id: 'total_performance_tracker',

        translations: {
            current_period_cost: 'Расход текущий период',
            current_period_revenue: 'Доход текущий период',
            current_period_profit: 'Прибыль текущий период',
            current_period_roi: 'ROI текущий период',
            current_period_clicks: 'Клики текущий период',
            current_period_leads: 'Лиды текущий период',
            previous_period_cost: 'Расход предыдущий период',
            previous_period_revenue: 'Доход предыдущий период',
            previous_period_profit: 'Прибыль предыдущий период',
            previous_period_roi: 'ROI предыдущий период',
            cost_change: 'Изменение расхода',
            revenue_change: 'Изменение дохода',
            profit_change: 'Изменение прибыли',
            roi_change: 'Изменение ROI',
            clicks_change: 'Изменение кликов',
            leads_change: 'Изменение лидов',
            trend: 'Тренд',
            total_cost: 'Общий расход',
            total_revenue: 'Общий доход',
            total_profit: 'Общая прибыль',
            roi: 'ROI',
            clicks: 'Клики',
            leads: 'Лиды',
            approved_leads: 'Одобренные лиды'
        },

        algorithm: `
            <ol>
                <li>Загружаются агрегированные метрики портфеля за выбранный период (по умолчанию 7 дней)</li>
                <li>Загружаются аналогичные данные за предыдущий период равной длины для сравнения</li>
                <li>Рассчитываются ключевые метрики текущего периода: общий расход, доход, прибыль, ROI, клики, лиды</li>
                <li>Рассчитываются аналогичные метрики предыдущего периода</li>
                <li>Определяются процентные изменения по каждой метрике между периодами</li>
                <li>Рассчитывается общий тренд на основе трех ключевых метрик (доход, прибыль, ROI)</li>
                <li>Тренд классифицируется как: improving (растущий), stable (стабильный), declining (падающий)</li>
                <li>Формируются алерты о тренде, ROI, расходах и прибыли</li>
            </ol>
        `,

        metrics: `
            <li><strong>Текущий период</strong> - все метрики за выбранный период анализа</li>
            <li><strong>Общий расход (Cost)</strong> - сумма расходов всех кампаний за период, $</li>
            <li><strong>Общий доход (Revenue)</strong> - сумма доходов всех кампаний за период, $</li>
            <li><strong>Общая прибыль (Profit)</strong> - разница между доходом и расходом, $</li>
            <li><strong>ROI портфеля</strong> - (Доход - Расход) / Расход * 100, %</li>
            <li><strong>Общие клики</strong> - сумма кликов всех кампаний за период</li>
            <li><strong>Общие лиды</strong> - сумма лидов всех кампаний за период</li>
            <li><strong>Предыдущий период</strong> - аналогичные метрики за предыдущий период</li>
            <li><strong>Изменения</strong> - процентные (%) изменения между периодами</li>
            <li><strong>Тренд</strong> - общая динамика: improving (улучшается), stable (стабилен), declining (ухудшается)</li>
        `,

        paramTranslations: {
            days: 'Период анализа',
            min_change_threshold: 'Порог изменения (%)',
            high_roi_threshold: 'Порог высокого ROI (%)',
            high_cost_change_threshold: 'Порог изменения расходов (%)'
        },

        /**
         * Отрисовка главного контента для total_performance_tracker
         */
        renderTable: function(results, container) {
            if (!results.data) {
                container.innerHTML = '<p class="text-muted">Нет данных для отображения</p>';
                return;
            }

            const current = results.data.current_period || {};
            const previous = results.data.previous_period || {};
            const changes = results.data.changes || {};
            const trend = results.data.trend || 'stable';
            const period = results.data.period || {};

            // Определяем цвета для тренда
            const trendColors = {
                'improving': '#28a745',
                'stable': '#6c757d',
                'declining': '#dc3545'
            };

            const trendLabels = {
                'improving': 'Улучшается',
                'stable': 'Стабилен',
                'declining': 'Ухудшается'
            };

            // Вспомогательная функция для форматирования изменений
            const formatChange = (value) => {
                if (value > 0) return `<span class="text-success">+${value.toFixed(1)}%</span>`;
                if (value < 0) return `<span class="text-danger">${value.toFixed(1)}%</span>`;
                return `<span class="text-muted">${value.toFixed(1)}%</span>`;
            };

            // Вспомогательная функция для форматирования ROI изменения
            const formatROIChange = (value) => {
                if (value > 0) return `<span class="text-success">+${value.toFixed(1)}%</span>`;
                if (value < 0) return `<span class="text-danger">${value.toFixed(1)}%</span>`;
                return `<span class="text-muted">${value.toFixed(1)}%</span>`;
            };

            // Создаем displaySummary из current_period для стандартной сетки
            const displaySummary = {
                total_cost: current.total_cost || 0,
                total_revenue: current.total_revenue || 0,
                total_profit: current.total_profit || 0,
                roi: current.total_roi || 0,
                clicks: current.total_clicks || 0,
                leads: current.total_leads || 0,
                trend: trend
            };

            let html = `
                <div class="performance-tracker-main">
                    <!-- Тренд индикатор -->
                    <div class="trend-indicator-banner" style="background: rgba(${
                        trend === 'improving' ? '40, 167, 69' :
                        trend === 'declining' ? '220, 53, 69' :
                        '108, 117, 125'
                    }, 0.1); border-left: 4px solid ${trendColors[trend]}; padding: 16px; margin-bottom: 20px; border-radius: 4px;">
                        <div style="display: flex; align-items: center; gap: 16px;">
                            <div>
                                <div class="text-muted" style="font-size: 12px; text-transform: uppercase; margin-bottom: 4px;">Общий тренд</div>
                                <div style="font-size: 20px; font-weight: bold; color: ${trendColors[trend]};">
                                    ${trendLabels[trend]}
                                </div>
                            </div>
                        </div>
                    </div>

                    <!-- Главные метрики в стандартной summary-grid -->
                    <div class="summary-grid">
                        <div class="summary-card">
                            <div class="summary-label">ROI</div>
                            <div class="summary-value ${displaySummary.roi >= 0 ? 'text-success' : 'text-danger'}">
                                ${displaySummary.roi.toFixed(1)}%
                            </div>
                            <div class="text-muted" style="font-size: 11px; margin-top: 4px;">
                                ${formatChange(changes.roi_change || 0)}
                            </div>
                        </div>

                        <div class="summary-card">
                            <div class="summary-label">Доход</div>
                            <div class="summary-value text-success">
                                ${formatCurrency(displaySummary.total_revenue)}
                            </div>
                            <div class="text-muted" style="font-size: 11px; margin-top: 4px;">
                                ${formatChange(changes.revenue_change || 0)}
                            </div>
                        </div>

                        <div class="summary-card">
                            <div class="summary-label">Расход</div>
                            <div class="summary-value">
                                ${formatCurrency(displaySummary.total_cost)}
                            </div>
                            <div class="text-muted" style="font-size: 11px; margin-top: 4px;">
                                ${formatChange(changes.cost_change || 0)}
                            </div>
                        </div>

                        <div class="summary-card">
                            <div class="summary-label">Прибыль</div>
                            <div class="summary-value ${displaySummary.total_profit >= 0 ? 'text-success' : 'text-danger'}">
                                ${formatCurrency(displaySummary.total_profit)}
                            </div>
                            <div class="text-muted" style="font-size: 11px; margin-top: 4px;">
                                ${formatChange(changes.profit_change || 0)}
                            </div>
                        </div>

                        <div class="summary-card">
                            <div class="summary-label">Клики</div>
                            <div class="summary-value">
                                ${displaySummary.clicks}
                            </div>
                            <div class="text-muted" style="font-size: 11px; margin-top: 4px;">
                                ${formatChange(changes.clicks_change || 0)}
                            </div>
                        </div>

                        <div class="summary-card">
                            <div class="summary-label">Лиды</div>
                            <div class="summary-value">
                                ${displaySummary.leads}
                            </div>
                            <div class="text-muted" style="font-size: 11px; margin-top: 4px;">
                                ${formatChange(changes.leads_change || 0)}
                            </div>
                        </div>
                    </div>

                    <!-- Сравнительная таблица -->
                    <div class="comparison-table">
                        <h3>Сравнение периодов</h3>
                        <div class="table-container">
                            <table>
                                <thead>
                                    <tr>
                                        <th>Метрика</th>
                                        <th>Текущий период</th>
                                        <th>Предыдущий период</th>
                                        <th>Изменение</th>
                                    </tr>
                                </thead>
                                <tbody>
                                    <tr>
                                        <td><strong>Расход</strong></td>
                                        <td>${formatCurrency(current.total_cost || 0)}</td>
                                        <td>${formatCurrency(previous.total_cost || 0)}</td>
                                        <td>${formatChange(changes.cost_change || 0)}</td>
                                    </tr>
                                    <tr>
                                        <td><strong>Доход</strong></td>
                                        <td>${formatCurrency(current.total_revenue || 0)}</td>
                                        <td>${formatCurrency(previous.total_revenue || 0)}</td>
                                        <td>${formatChange(changes.revenue_change || 0)}</td>
                                    </tr>
                                    <tr>
                                        <td><strong>Прибыль</strong></td>
                                        <td class="${current.total_profit >= 0 ? 'text-success' : 'text-danger'}" style="font-weight: bold;">
                                            ${formatCurrency(current.total_profit || 0)}
                                        </td>
                                        <td class="${previous.total_profit >= 0 ? 'text-success' : 'text-danger'}" style="font-weight: bold;">
                                            ${formatCurrency(previous.total_profit || 0)}
                                        </td>
                                        <td>${formatChange(changes.profit_change || 0)}</td>
                                    </tr>
                                    <tr>
                                        <td><strong>ROI</strong></td>
                                        <td>${current.total_roi !== undefined ? current.total_roi.toFixed(1) : 0}%</td>
                                        <td>${previous.total_roi !== undefined ? previous.total_roi.toFixed(1) : 0}%</td>
                                        <td>${formatROIChange(changes.roi_change || 0)}</td>
                                    </tr>
                                    <tr>
                                        <td><strong>Клики</strong></td>
                                        <td>${current.total_clicks !== undefined ? current.total_clicks : 0}</td>
                                        <td>${previous.total_clicks !== undefined ? previous.total_clicks : 0}</td>
                                        <td>${formatChange(changes.clicks_change || 0)}</td>
                                    </tr>
                                    <tr>
                                        <td><strong>Лиды</strong></td>
                                        <td>${current.total_leads !== undefined ? current.total_leads : 0}</td>
                                        <td>${previous.total_leads !== undefined ? previous.total_leads : 0}</td>
                                        <td>${formatChange(changes.leads_change || 0)}</td>
                                    </tr>
                                    <tr>
                                        <td><strong>Одобренные лиды</strong></td>
                                        <td>${current.total_a_leads !== undefined ? current.total_a_leads : 0}</td>
                                        <td>${previous.total_a_leads !== undefined ? previous.total_a_leads : 0}</td>
                                        <td>${formatChange(changes.a_leads_change || 0)}</td>
                                    </tr>
                                </tbody>
                            </table>
                        </div>
                    </div>

                    <!-- Info banner -->
                    <div class="info-banner">
                        <strong>Период:</strong> ${period.days || 7} дней (${period.date_from || 'N/A'} - ${period.date_to || 'N/A'})
                    </div>
                </div>
            `;

            container.innerHTML = html;
        }
    };

    // Регистрируем модуль
    if (typeof ModuleRegistry !== 'undefined') {
        ModuleRegistry.register(TotalPerformanceTrackerModule);
    }
})();
